Documentation/design-docs/standalone-gc-loading.md

```diff
@@ -0,0 +1,222 @@
+  incompatible interfaces. A change is considered breaking if it alters the semantics of an existing method or if 
+  it deletes or renames existing methods so that VTable layouts are not compatible.
+* If the EE's MinorVersion is greater than the MinorVersion obtained from the candidate GC, accept
+  (Forward compatability). The EE must take care not to call any new APIs that are not present in the version of
+  the candidate GC.
+* Otherwise, accept (Backward compatibility). It is perfectly safe to use a GC whose MinorVersion exceeds the EE's 
+  MinorVersion.
+
+The build version and name are not considered and are provided only for display/debug purposes.
+
+If this succeeds, the EE will transition to the next state in the loader state machine.
+
+### Initializing the GC
+
+Once the EE has verified that the version of the candidate GC is valid, it then proceeds to initialize the
+GC. It does so by loading (via `GetProcAddress`) and executing a function with this signature:
+
+```c++
+extern "C" HRESULT GC_Initialize(
+  /* In  */ IGCToCLR*,
+  /* Out */ IGCHeap**.
+  /* Out */ IGCHandleManager**,
+  /* Out */ GcDacVars*
+);
+```
+
+The EE will provide its implementation of `IGCToCLR` to the GC and the GC will provide its implementations of
+`IGCHeap`, `IGCHandleManager`, and `GcDacVars` to the EE. From here, if `GC_Initialize` returns a successful
+HRESULT, the GC is considered initialized and the remainder of EE startup continues. If `GC_Initialize` returns
+an error HRESULT, the initialization has failed.
+
+### Initializing the DAC
+
+The existence of a standalone GC is a debuggee process has implications for how the DAC is loaded and
+initializes itself. The DAC has access to implementation details of the GC that are not normally exposed as part
+of the `GC/EE` interfaces, and as such it is versioned differently.
+
+When the DAC is being initialized and it loads the `GcDacVars` structure from the debuggee process's memory, it
+must check the major and minor versions of the DAC, which are itself DAC variables exposed by a standalone GC.
+It then decides whether or not the loaded GC is compatible with the DAC that is currently executing. It does this
+in the same manner that the EE does:
+
+* If the major versions of the DAC and loaded GC do not agree, reject.
+* If the minor version of the DAC is greater than the minor version of the GC, accept but take care
+  not to invoke any new code paths not present in the target GC.
+* If the minor version of the DAC is less than or equal to the minor version of the GC, accept.
+
+If a DAC rejects a loaded GC, it will return `E_FAIL` from DAC APIs that would otherwise need to interact with the
+GC.
+
+### Loader State Machine
+
+The loading of a standalone GC by the EE is governed by a state machine with these states:
```

Comment:
The state machine makes sense if the set of transitions between different states is non-trivial. If the only way is from one state to the next with no forks or back edges, it is overkill to think about this as a state machine.

The sequence is described above. I do not think it needs to be repeated here. It is implementation detail.

Reply:
Makes sense. Perhaps it should be phrased more as a "status", like this: https://github.com/dotnet/coreclr/blob/2f37c02dfa3b649d4d7322724c95ab08cb12cb6f/src/vm/codeman.cpp#L1455. I was hoping for something where the point of failure is easily observed in a debugger or dump.